Implement UI process side of getFrame and getAllFrames

b-weinstein opened the pull request:
#### 69414c30375b5099697bf2b350d7c8f3576a5d58
<pre>
Implement UI process side of getFrame and getAllFrames
<a href="https://bugs.webkit.org/show_bug.cgi?id=264686">https://bugs.webkit.org/show_bug.cgi?id=264686</a>
<a href="https://rdar.apple.com/118278958">rdar://118278958</a>

Reviewed by Timothy Hatcher.

This patch sets up basic frame traversal APIs to:
1) Iterate all frames, adding their WebExtensionFrameParameters to a vector
2) Find a frame in a subtree

* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIWebNavigationCocoa.mm:
(WebKit::frameParametersForFrame): Set up the various parts of the WebExtensionFrameParameters object.
(WebKit::WebExtensionContext::webNavigationTraverseFrameTreeForFrame): Traverse the frame and subframes recursively.
(WebKit::WebExtensionContext::webNavigationFindFrameIdentifierInFrameTree): Find the frame with the given identifier recursively.
(WebKit::WebExtensionContext::webNavigationGetFrame): Call into webNavigationFindFrameIdentifierInFrameTree.
(WebKit::WebExtensionContext::webNavigationGetAllFrames): Call into webNavigationTraverseFrameTreeForFrame.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIWebNavigation.mm:
(TestWebKitAPI::TEST): Add tests for webNavigation.getAllFrames and webNavigation.getFrame (for both the main frame and subframes).

Canonical link: <a href="https://commits.webkit.org/270608@main">https://commits.webkit.org/270608@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/842cb981149dfe3ed73a2f93d25b97e412e430e9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25905 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4514 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27184 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28003 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23720 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/26221 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6277 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1942 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/23803 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26155 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3408 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22316 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28583 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3026 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23271 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/29334 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23642 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23648 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27209 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3049 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/1264 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4439 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6228 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3504 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3366 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIWebNavigationCocoa.mm:
```diff
@@ -32,13 +32,60 @@
+        false,
+
+        // url
+        extensionContext->hasPermission(frameURL, tab, { }) ? std::optional { frameURL } : std::nullopt,
```
`hasPermission(frameURL, tab)` will work, it defaults to empty.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIWebNavigationCocoa.mm:
```diff
@@ -32,13 +32,60 @@
+
+    for (_WKFrameTreeNode *childFrame in frame.childFrames) {
+        auto matchingChildFrame = webNavigationFindFrameIdentifierInFrameTree(childFrame, frame, tab, targetFrameIdentifier);
+        if (matchingChildFrame)
```
Can be one line:

```
if (auto matchingChildFrame = webNavigationFindFrameIdentifierInFrameTree(childFrame, frame, tab, targetFrameIdentifier))
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIWebNavigationCocoa.mm:
```diff
@@ -53,8 +100,9 @@
 
-    // FIXME: rdar://102820594 - Implement this.
-    completionHandler({ });
+    [webView _frames:makeBlockPtr([this, protectedThis = Ref { *this }, completionHandler = WTFMove(completionHandler), tab = WTFMove(tab), frameIdentifier = WTFMove(frameIdentifier)] (_WKFrameTreeNode *mainFrame) mutable {
```
```
makeBlockPtr([this, protectedThis = Ref { *this }, completionHandler = WTFMove(completionHandler), tab, frameIdentifier]
```

`tab` and `frameIdentifier` can just be copied since it is `RefPtr` and basically a `uint64_t`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIWebNavigationCocoa.mm:
```diff
@@ -71,8 +119,11 @@
 
-    // FIXME: rdar://102820594 - Implement this.
-    completionHandler({ });
+    [webView _frames:makeBlockPtr([this, protectedThis = Ref { *this }, completionHandler = WTFMove(completionHandler), tab = WTFMove(tab)] (_WKFrameTreeNode *mainFrame) mutable {
```
Ditto for copying `tab`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIWebNavigationCocoa.mm:
```diff
@@ -32,13 +32,60 @@
 
 namespace WebKit {
 
+static WebExtensionFrameParameters frameParametersForFrame(_WKFrameTreeNode *frame, _WKFrameTreeNode *parentFrame, WebExtensionTab* tab, WebExtensionContext* extensionContext, BOOL includeFrameIdentifier)
```
`bool` instead of `BOOL`.

b-weinstein replied:
I wasn't sure since this was an Obj-C++ file. I'll go with bool.

xeenon replied:
I'm pretty sure `BOOL` is `bool` in ObjC++, but just to be sure.

`bool` has better type conversion semantics, since returning a pointer will properly convert to 0 or 1, when compared to `BOOL` which typically requires `!!` for pointers to not be a ticking time bomb with the bit lottery.

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIWebNavigation.mm:
```diff
@@ -226,6 +226,168 @@
+        { "/frame.html"_s, { { { "Content-Type"_s, "text/html"_s } }, "<body style='background-color: blue'></body>"_s } },
+    }, TestWebKitAPI::HTTPServer::Protocol::Http);
+
+    auto *manifest = @{ @"manifest_version": @3, @"permissions": @[ @"webNavigation" ], @"background": @{ @"scripts": @[ @"background.js" ], @"type": @"module", @"persistent": @NO } };
```
This can move to a file static called `webNavigationManifest` so it can be used by other tests.

b-weinstein replied:
I'll create it and adopt it everywhere needed in this file.

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIWebNavigation.mm:
```diff
@@ -226,6 +226,168 @@
+        { "/frame.html"_s, { { { "Content-Type"_s, "text/html"_s } }, "<body style='background-color: blue'></body>"_s } },
+    }, TestWebKitAPI::HTTPServer::Protocol::Http);
+
+    auto *manifest = @{ @"manifest_version": @3, @"permissions": @[ @"webNavigation" ], @"background": @{ @"scripts": @[ @"background.js" ], @"type": @"module", @"persistent": @NO } };
```
Ditto on using a file static.

b-weinstein committed:
Implement UI process side of getFrame and getAllFrames
https://bugs.webkit.org/show_bug.cgi?id=264686
rdar://118278958

Reviewed by Timothy Hatcher.

This patch sets up basic frame traversal APIs to:
1) Iterate all frames, adding their WebExtensionFrameParameters to a vector
2) Find a frame in a subtree

* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIWebNavigationCocoa.mm:
(WebKit::frameParametersForFrame): Set up the various parts of the WebExtensionFrameParameters object.
(WebKit::WebExtensionContext::webNavigationTraverseFrameTreeForFrame): Traverse the frame and subframes recursively.
(WebKit::WebExtensionContext::webNavigationFindFrameIdentifierInFrameTree): Find the frame with the given identifier recursively.
(WebKit::WebExtensionContext::webNavigationGetFrame): Call into webNavigationFindFrameIdentifierInFrameTree.
(WebKit::WebExtensionContext::webNavigationGetAllFrames): Call into webNavigationTraverseFrameTreeForFrame.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIWebNavigation.mm:
(TestWebKitAPI::TEST): Add tests for webNavigation.getAllFrames and webNavigation.getFrame (for both the main frame and subframes).

Canonical link: https://commits.webkit.org/270608@main
